lib/basedriver/image-helpers.js

```diff
@@ -1,32 +1,69 @@
-import {PNG} from 'pngjs';
+import { Buffer } from 'buffer';
+import { PNG } from 'pngjs';
+import B from 'bluebird';
 
-function cropBase64Image (base64EncodedImage, rect) {
-  let pngImage = PNG.sync.read(Buffer.from(base64EncodedImage, 'base64'));
+async function createImageFromBase64 (base64EncodedImage) {
+  const buffer = Buffer.from(base64EncodedImage, 'base64');
+  return await createImageFromBuffer(buffer);
+}
+
+async function createImageFromBuffer (imageBuffer) {
+  return new B((resolve, reject) => {
+    const image = new PNG({filterType: 4});
+    image.parse(imageBuffer, (err, image) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      if (err) { return reject(err); }
+      resolve(image);
+    });
+  });
+}
+
+async function createBase64FromImage (image) {
+  return new B((resolve, reject) => {
+    const chunks = [];
+    image.pack().on('data', (chunk) => chunks.push(chunk)).on('end', () => {
+      resolve(Buffer.concat(chunks).toString('base64'));
+    }).on('error', (err) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      reject(err);
+    });
+  });
+}
 
-  if (rect.top < 0 || rect.top >= pngImage.height || rect.left < 0 || rect.left >= pngImage.width) {
+function cropImage (image, rect) {
+  const imageRect = {width: image.width, height: image.height};
+  const intersectionRect = getRectIntersection(rect, imageRect);
+  if (!isValidRect(intersectionRect)) {
     throw new Error('rect is outside the image bounds!');
   }
 
-  const yStart = rect.top,
-        yEnd = Math.min(rect.top + rect.height, pngImage.height),
-        xStart = rect.left,
-        xEnd = Math.min(rect.left + rect.width, pngImage.width);
-
-  let y, x, idx, i, croppedArray = [];
-  for (y = yStart; y < yEnd; y++) {
-    for (x = xStart; x < xEnd; x++) {
-      idx = (pngImage.width * y + x) << 2;
-      for (i = 0; i < 4; i++) {
-        croppedArray.push(pngImage.data[idx + i]);
+  const yEnd = Math.min(rect.top + rect.height, image.height);
+  const xEnd = Math.min(rect.left + rect.width, image.width);
+
+  let croppedArray = [];
+  for (let y = rect.top; y < yEnd; y++) {
+    for (let x = rect.left; x < xEnd; x++) {
+      const idx = (image.width * y + x) << 2;
```

Comment:
is it ok that we use _image.width_ to generate the data and then change it?

Reply:
It is ok because we override it only after the loop, but refactored to be more clear